### ISSUE TITLE
Fix L-Shelf Door Sizing and Configurations

### ISSUE THREAD

**Issue by TEDSTANDAGE** [CCDevNelson/Snap-Updates#169]
For L-Shelves (after https://github.com/CCDevNelson/Snap-Updates/issues/168 is done), there should be two options for doors--the default should be the two doors opening independently of one another, each opening outwardly. Only one door will have a handle and it will be on the inside part of the angle of the L. The door with the handle has to be opened by the customer first and then they reach behind the second door to open it. The SNaP user should be able to choose which door has the handle (L or R). Each of the doors will have normal hinge drilling. Sizing should match the following as it relates to width (height will be normal door height for the number of holes in the partitions:

![image](https://user-images.githubusercontent.com/61025894/91890890-faed2c00-ec44-11ea-981d-0fd8a1fd9ef2.png)

The second option for L-Shelf doors should be for Lazy Susan style doors, where the two doors open with each other, with one door connected to the partition and the other door, and the other door only connected to the door that is connected to the partition. The user should be able to choose which direction the doors open. The door that is connected to the partition will have both sets of hinge drilling on it, while the door that is only connected to the other door does not have any hinge drilling. Sizing should match the following as it relates to width (again height is determined by the partition height):

![image](https://user-images.githubusercontent.com/61025894/91891257-91b9e880-ec45-11ea-8213-b01280bc8ce1.png)

**Comment by ClassyTeddy** [comment]
I believe this is ready for testing on the `Improving-L-and-Corner-Shelves` branch.

**Comment by sarahhightower** [comment]
This is such an awesome feature. Thanks for working on this. Below are my comments:

- The prompt for L-Shelves still says "Corner Shelf Prompt" (See visual 1 below)

- Is it possible to have the door styles by each other in the prompt? So for example, once the user selects that they want a door, the type of door (lazy Susan / double door) appear indented underneath as the options? 

- Is it possible for the door pull height to adjust with the door size? Or would we be able to give the same door options that appear in our standard door prompt (tall/base/upper)?

- In a double door test I selected left swing and the door with the handle appears to open right. (See visual 2 below) I think the door with the handle should represent the swing type, but correct me if I'm wrong @TylerStandage @TedCC1209. 

- Is it possible to see these doors open? 

- The 2D views lack some necessary information to reflect the L Shelves and the door styles. I'm guessing this might be something we will discuss in issue #281 though, so I won't go into detail unless requested.  

![1](https://user-images.githubusercontent.com/72047269/109332728-c20cc900-781b-11eb-8964-d6222bc70115.jpg)

![2](https://user-images.githubusercontent.com/72047269/109332732-c46f2300-781b-11eb-8386-30d41d5a487a.jpg)

**Comment by ClassyTeddy** [comment]
> Is it possible to have the door styles by each other in the prompt? So for example, once the user selects that they want a door, the type of door (lazy Susan / double door) appear indented underneath as the options?

This should be possible, would you want it like a dropdown combobox, or like 2 checkboxes and when you turn one on, it turns the other one off?

>Is it possible for the door pull height to adjust with the door size? Or would we be able to give the same door options that appear in our standard door prompt (tall/base/upper)

This is also possible if it is how we want it to function

> Is it possible to see these doors open?

This one I am far more unsure. It should be possible to see double doors open, but I am unsure how I would make the Lazy Susan Doors open properly. I think I might be able to have them open in a way where they stay at that 90 degree angle and open, but I don't know if I can make it "fold" like I believe it should.

> In a double door test I selected left swing and the door with the handle appears to open right. (See visual 2 below) I think the door with the handle should represent the swing type, but correct me if I'm wrong

I had noticed this as well, and I am unsure if we want the wording to be `Left Swing` but instead want it as a combobox option with 2 options, `Pull on Left Door` and `Pull on Right Door`.

**Comment by sarahhightower** [comment]
1. Without seeing how this would look I think I like the idea of a dropdown. Open to others feedback.
2. @TedCC1209 @TylerStandage what do you think?
3.  The lazy susan door option you outlined sounds fine to me.
4.  The option you proposed sounds good to me.

**Comment by TedCC1209** [comment]
> 1. Without seeing how this would look I think I like the idea of a dropdown. Open to others feedback.

I like the idea of a dropdown as well. Can we have the options be called "Reach Back" and "Lazy Susan" doors?

> 2. @TedCC1209 @TylerStandage what do you think?

Yes that would be fine to have it function as you have proposed it.

> 3. The lazy susan door option you outlined sounds fine to me.

It's a good start. The reach back doors may give you a little trouble too because the door with the handle will have to open first and then the reach back door, not sure how that interplay would work.

> 4. The option you proposed sounds good to me.

I agree that yes we should change that wording to "Pull on left door" or "Pull on right door" which would also control the swing for Lazy Susan doors.

We also need to get rid of the "Force Double Doors" for L-Shelves (not corner cabinets).

**Comment by ClassyTeddy** [comment]
This is mostly ready for testing on the `Improving-L-and-Corner-Shelves` branch, except for the L Shelf door opening, which I go into more detail on issue #170, as well as the following comment:

> Is it possible for the door pull height to adjust with the door size? Or would we be able to give the same door options that appear in our standard door prompt (tall/base/upper)

So how do we want this to function? Should it adjust automatically with door size, based on a certain distance from the top or bottom of the door, and the pull location merely moves it up and down from there? Or do we want it to have the options similar to regular doors now?

**Comment by TedCC1209** [comment]
I would say let's have the same door options as regular doors to determine the handle placement (upper, base, tall). That way doors would be somewhat uniform across the software.

**Comment by TylerStandage** [comment]
I agree.  Makes the software a bit more intuitive, especially for newer designers, if similar items/options are treated the same way in different scenarios.

**Comment by TedCC1209** [comment]
@ClassyTeddy I am wondering if you put hinge drilling in for the doors (particularly on the inside lazy susan door where there will be hinge drilling for connecting to the partition as well as hinge drilling for connecting to the other door which will not have any drilling).

**Comment by ClassyTeddy** [comment]
There is currently no drilling for hinges on Corner or L Shelf doors I believe. I will have to double check.

**Comment by ClassyTeddy** [comment]
The Pull Location Changes are ready for testing on the `Improving-L-and-Corner-Shelves` branch. I also changed some of the UI Slightly because 3 full combo boxes were a little large. Please let me know your thoughts on this.

**Comment by TylerStandage** [comment]
The door swings looks really good... great job! A few comments:

The prompt still says Corner Shelves Prompt:
![image](https://user-images.githubusercontent.com/77167318/114627533-40a5b480-9c6a-11eb-8bb3-59896e5f3363.png)

Having the handle go to a set height based on pull type can yield some weird results. For example, the default, even for L-shelves that have a toe kick and go to the floor, puts the handle near the floor. If you then change the height of the pards to be uppers and hanging, the handle remains on the bottom. Take a look and see if there are easy fixes to this. If not, we won't worry about it for this release since the designer can still adjust to wherever they need the pull to be located.

**Comment by ClassyTeddy** [comment]
> For example, the default, even for L-shelves that have a toe kick and go to the floor, puts the handle near the floor. If you then change the height of the pards to be uppers and hanging, the handle remains on the bottom.

I can take a look at this, but isn't that how the base setting for pulls intended to work? I do want to say that this is how the pulls for doors work, and the different pull types affect basically the starting point for the pull distance. 

Base currently starts at 2 inches and measures from the bottom, meaning that the pulls center starts 2 inches from the base of the door, and it ignores the toe kick.

Tall currently starts at 40 inches and measures down from the top, meaning the pull starts 40 inches down from the top of the door. If the top of the door doesn't move (i.e. if it is made into an upper and it's height changes), then the pull won't move either. It will always stay so and so many inches from the top.

Upper functions exactly like tall, except it starts at a value of 2 inches. 

If we want to change how any of these settings affect pulls, let me know. But we may then want to consider changing the logic for doors as well, as this is exactly how then function for them.

I'd like @sarahhightower input as well, to see if this was what she had in mind.

**Comment by TylerStandage** [comment]
Hi Teddy, when I started with a tall unit with a toe kick, the default was to have the pull near the floor.  Then when I changed the assembly to be hanging and an upper, the door pull stayed very close to the floor (didn't move with the change in the assembly). This is probably an issue across the board that can be handled separately.  Perhaps we could make the default for any of these doors in assemblies with toe kicks to be the "tall" location for the pull rather than "base"?  @sarahhightower thoughts?

**Comment by sarahhightower** [comment]
@ClassyTeddy I think the base and upper handle placements are inverted. Base cabinets should have the handles 2" from the top. Upper cabinets should have the handle placed 2" from the bottom. This is how it functions with standard doors. Visual below.

![1](https://user-images.githubusercontent.com/72047269/114751649-4521aa00-9d0a-11eb-83a5-0ff02b9da8fa.jpg)

![2](https://user-images.githubusercontent.com/72047269/114751936-a184c980-9d0a-11eb-9c33-cf0a9f522557.jpg)

@TylerStandage with standard doors you have the option to choose the door hole size, so when you choose the type of cabinet the doors and handles shift to be that type of cabinet. For example, in my visuals above I chose 21H doors so when I chose base cabinets they went to the bottom and when I chose uppers they went to the top. Choosing a door size doesn't exist with L shelves so I think that's the reason that the door handle goes rouge when you choose a different type of cabinet. I think we should probably add the option to  choose the door size, with the default being full pard sized doors as I don't think it's very common to have both doors and shelves. But, this would make the prompts consistent and give designers that option should they want to do it. As far as making the default be tall doors when the cabinet has toe kicks, I think that would be good as well.
